src/jit/importer.cpp

```diff
@@ -10624,11 +10624,22 @@ void Compiler::impImportBlockCode(BasicBlock* block)
             SPILL_APPEND:
 
                 // We need to call impSpillLclRefs() for a struct type lclVar.
-                // This is done for non-block assignments in the handling of stloc.
-                if ((op1->OperGet() == GT_ASG) && varTypeIsStruct(op1->gtOp.gtOp1) &&
-                    (op1->gtOp.gtOp1->gtOper == GT_LCL_VAR))
+                if ((op1->OperGet() == GT_ASG) && varTypeIsStruct(op1->gtGetOp1()))
                 {
-                    impSpillLclRefs(op1->gtOp.gtOp1->AsLclVarCommon()->gtLclNum);
+                    GenTree*             lhs    = op1->gtGetOp1();
+                    GenTreeLclVarCommon* lclVar = nullptr;
+                    if (lhs->gtOper == GT_LCL_VAR)
+                    {
+                        lclVar = lhs->AsLclVarCommon();
+                    }
+                    else if (lhs->OperIsBlk())
+                    {
+                        lclVar = lhs->AsBlk()->Addr()->IsLocalAddrExpr();
```

Comment:
Is `IsLocalAddrExpr` really the right function to use here? There are a few that attempt to detect local addresses and they differ in the trees that they recognize. This one recognizes something like `ADD(ADDR(LCL_VAR, 2)` but that's less common during import. There's also `impIsAddressInLocal` that does something similar but instead of `ADD` nodes it looks for `FIELD` nodes, which of course are more common during import.

Reply:
Yes, I think you're right. `IsLocalAddrExpr()` would be the right thing to call after morph, but at this point we might see a `FIELD`. I'm going to try to get a repro case for that.

Reply:
Having spent some time on this, I'm fairly certain that this only needs to handle full assignments. The `FIELD` case is handled here: https://github.com/dotnet/coreclr/blob/master/src/jit/importer.cpp#L14507, and explicit `stloc` cases here: https://github.com/dotnet/coreclr/blob/master/src/jit/importer.cpp#L11103
.In the failure case the "assignment" is an `initobj`. The block cases handle this with a `goto SPILL_APPEND` which lands here. Pretty much anything else that would modify a local would involve explicitly taking its address (not just in the context of doing a field store).

I gave some thought to whether an assert could be added, but I was unable to do so. So, I'm actually going to simplify this check.